### PR TITLE
First attempt to membership demo.

### PR DESCRIPTION
**Pull request by sotr3**
The purpose is to discuss if including MembershipState for the two participants of a Transfer Contract is what we would like to have. If this is the case the code is a doing that.

### CODE REVIEW

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/contracts/SampleContract.kt [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import net.corda.core.contracts.TypeOnlyCommandData
+import net.corda.core.contracts.requireThat
+import net.corda.core.identity.Party
+import net.corda.core.transactions.LedgerTransaction
+import java.lang.IllegalArgumentException
+
+class SampleContract : Contract {
+    companion object {
+        // amount of billing chips to be paid for each transaction by its participants
+        //const val BILLING_CHIPS_TO_PAY = 10L
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/contracts/SampleContract.kt]: What are these comments related to?

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+class RequestMembershipForTransaction (val bno : Party)
+
+class IssueAssetFlow (val bno : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // query the vault with bno
+        val memberships= subFlow(GetMembershipsFlow(bno))
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: Space before "=". In intellij you can press Cmd + Shift + L to automatically reformat code

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        val ourMembership = memberships[ourIdentity]!!
+
+        val builder = TransactionBuilder(notary)
+                .addOutputState(SampleState(ourIdentity), SampleContract.CONTRACT_NAME)
+                .addCommand(SampleContract.Commands.Issue(), ourIdentity.owningKey)
+                .addReferenceState(ReferencedStateAndRef(ourMembership))
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: The contract should check that the reference state is indeed inside the transaction

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        if (ourIdentity != sampleState.state.data.owner)
+            throw FlowException("Only owner of the state can transfer it")
+        if (ourIdentity == partyToTransferTo)
+            throw FlowException("sender and recipient should be different parties")
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // get the membership from the node's membership cache
+        val membershipsCacheHolderService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
+        val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity)
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: Would make sense to use the same API across the whole app. I.e. either `MembershipsCacheHolder` or `GetMembershipFlow`

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+class IssueAssetFlow (val bno : Party) : FlowLogic<SignedTransaction>() {
+    @Suspendable
+    override fun call() : SignedTransaction {
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // query the vault with bno
+        val memberships= subFlow(GetMembershipsFlow(bno))
+        val ourMembership = memberships[ourIdentity]!!
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: This will throw `NullPointerException` if membership is missing, which isn't very user friendly. In kotlin you can use the following syntax `val ourMembership = membership[ourIdentity] ?: throw FlowException("Membership for $ourIdentitiy has not been found")`

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        if (ourIdentity == partyToTransferTo)
+            throw FlowException("sender and recipient should be different parties")
+        // for the sake of demo we are just taking the first available notary
+        val notary = serviceHub.networkMapCache.notaryIdentities.single()
+
+        // get the membership from the node's membership cache
+        val membershipsCacheHolderService = serviceHub.cordaService(MembershipsCacheHolder::class.java)
+        val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity)
+
+        if (ourMembership == null) {
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: In kotlin you can use a shorted syntax for that `val ourMembership = membershipsCacheHolderService.cache.getMembership(bno, ourIdentity) ?: throw FlowException("Invalid membership: ourMembership is null")`

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 4 above)
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: We should also verify that `partyToTransferTo` has membership as well

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                .addOutputState(SampleState(partyToTransferTo))
+                .addCommand(SampleContract.Commands.Transfer(), ourIdentity.owningKey, partyToTransferTo.owningKey)
+
+        // create a session
+        val counterPartySession = initiateFlow(partyToTransferTo)
+        // requesting other party to send us their billing states
+        counterPartySession.send(RequestMembershipForTransaction(bno))
+
+
+        val counterPartyMembershipState = subFlow(ReceiveStateAndRefFlow<MembershipState<Any>>(counterPartySession)).single()
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: Why not to get this state directly from the memberships cache?

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    }
+}
+
+@InitiatedBy(TransferAssetFlow::class)
+class TransferAssetFlowResponder(val session : FlowSession) : FlowLogic<Unit>() {
+    @Suspendable
+    override fun call() {
+
+        // receiving request for billing state
+        val bno = session.receive<RequestMembershipForTransaction>().unwrap { it }
```
Comment (ivanr3) [inline bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: It's important to always verify things that you are received from the wire. For example, should we verify that the received identity is indeed a legit BNO?

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/contracts/SampleContract.kt (hunk 1 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/contracts/SampleContract.kt]: oops. Fixed.

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 2 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: nice. Done

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 3 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: In the Commands.Issue I added the following check:
"There should be one reference input state of type MembershipState" using (tx.referenceInputsOfType<MembershipState<Any>>().size == 1)

In the Commands.Transfer I added the following check:
"There should be two reference input state of type MembershipState" using (tx.referenceInputsOfType<MembershipState<Any>>().size == 2)

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 4 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: Used GetMembershipFlow everywhere.
Done.

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 5 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: nice. done

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 6 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: done!

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 4 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: @ivanr3 
Right now, in the TransferAssetFlow this is checked in AssetContract.Commands.Transfer() when the verify() runs.
The membership for the counterparty is obtained for the counterparty as follows (without the ReceiveStateAndRefFlow the chain resolution fails - as it should):
        // create a session
        val counterPartySession = initiateFlow(partyToTransferTo)
        // requesting other party to send us their membership state
        counterPartySession.send(RequestMembershipForTransaction(bno))
        val counterPartyMembershipState =
                subFlow(ReceiveStateAndRefFlow<MembershipState<Any>>(counterPartySession)).single()

Question: On top of the check above, should the membership of the counterparty be checked in the TransferAssetFlow as follows : subFlow(GetMembershipsFlow(bno))[partyToTransferTo]
Is that what you have in mind?

Am I missing sth?

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 7 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: See above. That's what I tried to do initially. But if you do that the transaction resolution fails because you do not have the chain for the counterparty's membership.

File: bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt (hunk 8 above)
Comment (sotr3) [reply bn-apps/memberships-management/membership-management-demo/src/main/kotlin/com/r3/businessnetworks/memberships/demo/flows/Flows.kt]: good point. done

**Comment by sotr3** [comment]
@ivanr3  I have applied the comments above and/or added my own comments.
Next thing to do is add Cash.
It's not finished yet. Planning to have a separate commit for that.

**Comment by ivanr3** [comment]
Hi @sotr3 . THis PR seems to be stale. Closing for now.
